modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -1,67 +1,112 @@
-            }
-        }
-      }.flatten
+    globalSnapshotLocalFileStorage: GlobalSnapshotLocalFileStorage[F],
+    genesis: GlobalSnapshot,
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorage[F]] = {
+    def mkHeadRef = Ref.of[F, GlobalSnapshot](genesis)
+    def mkCache = MapRef.ofSingleImmutableMap[F, SnapshotOrdinal, GlobalSnapshot](Map.empty)
+    def mkOffloadQueue = Queue.unbounded[F, SnapshotOrdinal]
 
-    def get(ordinal: SnapshotOrdinal): F[Option[GlobalSnapshot]] =
-      snapshotsRef.get.map(_.find(_.ordinal === ordinal))
+    def mkLogger = Slf4jLogger.create[F]
 
-    def getLast: F[GlobalSnapshot] = snapshotsRef.get.map(_.head)
+    mkLogger.flatMap { implicit logger =>
+      (mkHeadRef, mkCache, mkOffloadQueue).mapN {
+        make(_, _, _, globalSnapshotLocalFileStorage, inMemoryCapacity)
+      }.flatten
+    }
+  }
 
-    def getStateChannelSnapshotUntilOrdinal(
-      ordinal: SnapshotOrdinal
-    )(address: Address): F[Option[StateChannelSnapshotBinary]] =
-      snapshotsRef.get.map { snapshots =>
-        snapshots.find { snapshot =>
-          snapshot.ordinal <= ordinal && snapshot.stateChannelSnapshots.contains(address)
-        }.flatMap { snapshot =>
-          snapshot.stateChannelSnapshots.get(address).map(_.head)
+  def make[F[_]: Async: Logger: KryoSerializer](
+    headRef: Ref[F, GlobalSnapshot],
+    cache: MapRef[F, SnapshotOrdinal, Option[GlobalSnapshot]],
+    offloadQueue: Queue[F, SnapshotOrdinal],
+    globalSnapshotLocalFileStorage: GlobalSnapshotLocalFileStorage[F],
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorage[F]] = {
+
+    def offloadProcess =
+      Stream
+        .fromQueueUnterminated(offloadQueue)
+        .evalTap { ordinal =>
+          cache(ordinal).get.flatMap {
+            case Some(snapshot) =>
+              globalSnapshotLocalFileStorage.write(snapshot).rethrowT.handleErrorWith { e =>
+                Logger[F].error(e)(s"Failed writing global snapshot to disk! Snapshot ordinal=${snapshot.ordinal.show}")
+              } >> cache(ordinal).set(none)
+            case None => MonadThrow[F].raiseError[Unit](new Throwable("Unexpected state!"))
+          }
         }
+        .compile
+        .drain
+
+    Spawn[F].start { offloadProcess }.flatMap { _ =>
+      headRef.get.flatMap { h =>
+        cache(h.ordinal).set(h.some)
       }
-  }
+    }.map { _ =>
+      new GlobalSnapshotStorage[F] {
+        def prepend(snapshot: GlobalSnapshot) =
+          headRef.modify { current =>
+            isNextSnapshot(current, snapshot) match {
+              case Left(e) =>
+                (current, Logger[F].error(e)(s"Unexpected error while checking next snapshot chain!").map(_ => false))
+              case Right(isNext) if isNext =>
+                def run =
+                  cache(snapshot.ordinal).set(snapshot.some) >>
+                    snapshot.ordinal
+                      .partialPreviousN(NonNegLong.unsafeFrom(inMemoryCapacity.toLong))
```

Comment:
I think we shouldn't use `unsafe` on refined (maybe tests could be an exception to the rule).

Reply:
`inMemoryCapacity` is a positive long, so it's safe to convert it into non-negative long. unfortunately, there is no embedded function that will convert from more strict to more general type